# Data Processing - Assignment 5

### Task 1 (20 points)

Here is a generic code to download a dataset from http://archive.ics.uci.edu/ml/index.php and run a regression task.

In [1]:
import time
import numpy as np
import pandas as pd
from sklearn import linear_model
from sklearn import datasets
from scipy.stats import describe
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import OneHotEncoder

Read about information on the bike-sharing dataset:

http://archive.ics.uci.edu/ml/datasets/bike+sharing+dataset

download and run this code on it.

bikeDataSet = genfromtxt('hour.csv', delimiter=',')

X = bikeDataSet[1:,2:-1]

y = bikeDataSet[1:,-1]

from sklearn import linear_model

estimator = linear_model.LinearRegression()

score = cross_val_score(estimator, X, y).mean()

print(f"Score = {score:.3f}.")

Try putting random features with 𝑋=𝑛𝑝.𝑟𝑎𝑛𝑑𝑜𝑚.𝑟𝑎𝑛𝑑(𝑋.𝑠ℎ𝑎𝑝𝑒[0],4)'' and see how the score changes. Also try the following estimator instead: 𝑙𝑖𝑛𝑒𝑎𝑟_𝑚𝑜𝑑𝑒𝑙.𝐿𝑎𝑠𝑠𝑜(𝑎𝑙𝑝ℎ𝑎=0.1)'' and report the result in your NoteBook

In [2]:
bikeDataSet = np.genfromtxt('/Users/varsha/Downloads/Bike-Sharing-Dataset/hour.csv', delimiter=',')
X = bikeDataSet[1:,2:-1]
y = bikeDataSet[1:,-1]
estimator = linear_model.LinearRegression()
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = 1.000.


Changing X value:

In [3]:
X = np.random.rand(X.shape[0],4)
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = -0.281.


In [4]:
# Experimenting with various x values
X = X/100
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = -0.281.


In [5]:
# Experimenting with various x values
X = np.random.rand(X.shape[0],1500)
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = -0.475.


Changing estimator to Lasso:

In [6]:
X = bikeDataSet[1:,2:-1]
estimator = linear_model.Lasso(alpha=0.1)
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = 1.000.


In [7]:
X = np.random.rand(X.shape[0],4)
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = -0.280.


We can see the change in score whenever we assign X value to np.random.rand(X.shape[0],4) where the score is reduced from 1 to -0.280. 

However, the score remains to be the same as LinearRegression when we assign linear_model.Lasso(alpha=0.1) as the estimator.

***

### Task 2 (20 points)

Download the full kddcup dataset from UCI repository and find out what each feature is about and which ones are categorical. Try to encode all the features using One hot encoding. Report how long the representation length is after the transformation. 

Use the skills you learned in the past modules to normalize the features so that each column has mean zero and unit standard deviation. 

Report how long it takes for your system to perform one-hot encoding and normalization together.

In [8]:
kddcupDataSet = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/kddcup99-mld/kddcup.data.gz")
kddcupDataSet.columns = ['duration', 'protocol_type', 'service', 'flag', 'src_bytes', 'dst_bytes', 'land', 'wrong_fragment', 'urgent', 'hot','num_failed_logins','logged_in','num_compromised','root_shell','su_attempted','num_root','num_file_creations','num_shells','num_access_files','num_outbound_cmds','is_hot_login','is_guest_login','count','srv_count','serror_rate','srv_serror_rate','rerror_rate','srv_rerror_rate','same_srv_rate','diff_srv_rate','srv_diff_host_rate','dst_host_count','dst_host_srv_count','dst_host_same_srv_rate','dst_host_diff_srv_rate','dst_host_same_src_port_rate','dst_host_srv_diff_host_rate','dst_host_serror_rate','dst_host_srv_serror_rate','dst_host_rerror_rate','dst_host_srv_rerror_rate','connection_type']

In [9]:
kddcupDataSet.head()

duration protocol_type service flag  src_bytes  dst_bytes  land  \
0         0           tcp    http   SF        162       4528     0   
1         0           tcp    http   SF        236       1228     0   
2         0           tcp    http   SF        233       2032     0   
3         0           tcp    http   SF        239        486     0   
4         0           tcp    http   SF        238       1282     0   

   wrong_fragment  urgent  hot  ...  dst_host_srv_count  \
0               0       0    0  ...                   1   
1               0       0    0  ...                   2   
2               0       0    0  ...                   3   
3               0       0    0  ...                   4   
4               0       0    0  ...                   5   

   dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                     1.0                     0.0   
1                     1.0                     0.0   
2                     1.0                     0.0   
3                     1.0                     0.0   
4                     1.0                     0.0   

   dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                         1.00                          0.0   
1                         0.50                          0.0   
2                         0.33                          0.0   
3                         0.25                          0.0   
4                         0.20                          0.0   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                   0.0                       0.0                   0.0   
1                   0.0                       0.0                   0.0   
2                   0.0                       0.0                   0.0   
3                   0.0                       0.0                   0.0   
4                   0.0                       0.0                   0.0   

   dst_host_srv_rerror_rate  connection_type  
0                       0.0          normal.  
1                       0.0          normal.  
2                       0.0          normal.  
3                       0.0          normal.  
4                       0.0          normal.  

[5 rows x 42 columns]

In [10]:
kddcupDataSet.dtypes

duration                         int64
protocol_type                   object
service                         object
flag                            object
src_bytes                        int64
dst_bytes                        int64
land                             int64
wrong_fragment                   int64
urgent                           int64
hot                              int64
num_failed_logins                int64
logged_in                        int64
num_compromised                  int64
root_shell                       int64
su_attempted                     int64
num_root                         int64
num_file_creations               int64
num_shells                       int64
num_access_files                 int64
num_outbound_cmds                int64
is_hot_login                     int64
is_guest_login                   int64
count                            int64
srv_count                        int64
serror_rate                    float64
srv_serror_rate          

In [11]:
start_time = time.time()
category = kddcupDataSet.select_dtypes(include='object')
category_cols = category.columns.tolist()
numerical = kddcupDataSet.select_dtypes(exclude='object')
numerical_cols = numerical.columns.tolist()

transformation = OneHotEncoder().fit_transform(kddcupDataSet[category_cols[:-1]])
normalization = (kddcupDataSet[numerical_cols]-kddcupDataSet[numerical_cols].mean())/kddcupDataSet[numerical_cols].std()
print ("Time taken for one-hot encoding and normalization: ",time.time()-start_time)

Time taken for one-hot encoding and normalization:  8.147058963775635


***
### Task 3 (10 points)

Run the following code with your guesses of optimal alpha, and report the value that gave the highest score:

from sklearn import datasets

from scipy.stats import describe

from sklearn.svm import SVC

alpha = 1

iris = datasets.load_iris()

X = iris.data

y = iris.target

X = X * alpha

estimator = SVC(kernel = 'linear')

score = cross_val_score(estimator, X, y).mean()

print(f"Score = {score:.3f}.")


In [12]:
alpha = -.99
iris = datasets.load_iris()
X = iris.data
y = iris.target
X = X * alpha
estimator = SVC(kernel = 'linear')
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = 0.987.


In [13]:
alpha = .8
iris = datasets.load_iris()
X = iris.data
y = iris.target
X = X * alpha
estimator = SVC(kernel = 'linear')
score = cross_val_score(estimator, X, y).mean()
print(f"Score = {score:.3f}.")

Score = 0.987.


I tried out various numbers ranging from 0-1000 and also negative values. The highest score obtained was 0.987 with alpha value being 0.8 and -0.99.